File: tasks_js_solution/tasks.ts
export type TaskStatus = "pending" | "completed" | "failed";

export interface Task {
  type: string;
  status: TaskStatus;
  destination?: string;
  payload?: string;
}

// Interfaz clave para el polimorfismo
export interface TaskHandler {
  execute(task: Task): Promise<void>;
}

// Implementaciones (SRP)
export class EmailHandler implements TaskHandler {
  constructor(private readonly smtpClient?: any) {}

  async execute(task: Task): Promise<void> {
    console.log("Conectando al servidor SMTP...");
    console.log(
      `Enviando correo a ${task.destination} con el mensaje: ${task.payload}`,
    );
  }
}

export class ReportHandler implements TaskHandler {
  constructor(private readonly dbConnection?: any) {}

  async execute(task: Task): Promise<void> {
    console.log("Conectando a la base de datos...");
    console.log(`Generando reporte PDF para ${task.destination}`);
  }
}

export class WhatsAppHandler implements TaskHandler {
  constructor(private readonly whatsappClient?: any) {}

  async execute(task: Task): Promise<void> {
    console.log("Conectando a WhatsApp API...");
    console.log(`Enviando WhatsApp a ${task.destination}: ${task.payload}`);
  }
}

export class SMSHandler implements TaskHandler {
  constructor(private readonly smsClient?: any) {}

  async execute(task: Task): Promise<void> {
    console.log("Conectando a SMS Gateway...");
    console.log(`Enviando SMS a ${task.destination}: ${task.payload}`);
  }
}

export class PushHandler implements TaskHandler {
  constructor(private readonly pushClient?: any) {}

  async execute(task: Task): Promise<void> {
    console.log("Conectando a Push Notification Service...");
    console.log(`Enviando push a ${task.destination}: ${task.payload}`);
  }
}

// El Orquestador (OCP)
export class TaskProcessor {
  constructor(private readonly handlers: Map<string, TaskHandler>) {}

  public async processAndNotify(tasks: Task[]): Promise<Task[]> {
    const results: Task[] = [];

    for (const task of tasks) {
      if (task.status !== "pending") continue;

      const handler = this.handlers.get(task.type);

      if (!handler) {
        console.warn(`No handler found for type: ${task.type}`);
        task.status = "failed";
        results.push(task);
        continue;
      }

      try {
        await handler.execute(task);
        task.status = "completed";
      } catch (error) {
        console.error(`Error processing task ${task.type}:`, error);
        task.status = "failed";
      }

      results.push(task);
    }

    return results;
  }
}

File: tasks_py_solution/tasks.py
from typing import Protocol, Dict, List, Optional
from dataclasses import dataclass
import logging

logging.basicConfig(level=logging.INFO)


@dataclass
class Task:
    type: str
    status: str
    destination: Optional[str] = None
    payload: Optional[str] = None


# 1. Definimos el contrato (Interfaz) para cualquier manejador de tareas
class TaskHandler(Protocol):
    def execute(self, task: Task) -> None:
        """Ejecuta la lógica específica de la tarea. Lanza una excepción si falla."""
        pass


# 2. Implementaciones concretas (Cumplen con Single Responsibility)
class EmailHandler:
    def __init__(self, smtp_client=None):
        self._smtp_client = smtp_client  # Inyección de dependencia

    def execute(self, task: Task) -> None:
        logging.info("Conectando al servidor SMTP...")
        logging.info(f"Enviando correo a {task.destination} con el mensaje: {task.payload}")


class ReportHandler:
    def __init__(self, db_connection=None):
        self._db_connection = db_connection

    def execute(self, task: Task) -> None:
        logging.info("Conectando a la base de datos...")
        logging.info(f"Generando reporte PDF para {task.destination}")


class WhatsAppHandler:
    def __init__(self, whatsapp_client=None):
        self._whatsapp_client = whatsapp_client

    def execute(self, task: Task) -> None:
        logging.info("Conectando a WhatsApp API...")
        logging.info(f"Enviando WhatsApp a {task.destination}: {task.payload}")


class SMSHandler:
    def __init__(self, sms_client=None):
        self._sms_client = sms_client

    def execute(self, task: Task) -> None:
        logging.info("Conectando a SMS Gateway...")
        logging.info(f"Enviando SMS a {task.destination}: {task.payload}")


class PushHandler:
    def __init__(self, push_client=None):
        self._push_client = push_client

    def execute(self, task: Task) -> None:
        logging.info("Conectando a Push Notification Service...")
        logging.info(f"Enviando push a {task.destination}: {task.payload}")


# 3. El Orquestador (Cumple con Open/Closed Principle)
class TaskProcessor:
    def __init__(self, handlers: Dict[str, TaskHandler]):
        # Recibe el registro de handlers por el constructor
        self._handlers = handlers

    def process_and_notify(self, tasks: List[Task]) -> List[Task]:
        results = []
        for task in tasks:
            if task.status != "pending":
                continue

            handler = self._handlers.get(task.type)

            if not handler:
                logging.warning(f"Handler no encontrado para el tipo: {task.type}")
                task.status = "failed"
                results.append(task)
                continue

            try:
                handler.execute(task)
                task.status = "completed"
            except Exception as e:
                logging.error(f"Error procesando tarea {task.type}: {e}")
                task.status = "failed"

            results.append(task)

        return results
